### DISCUSSION TITLE
How is ..HttpFoundation\Request available through dependency injection?

### DISCUSSION THREAD

**Discussion by jdevinemt** [symfony/symfony#44517] [Q&A] [answered]
I'm currently learning symfony and migrating an existing project into the framework (v5.4). I was looking into the best way to handle redirects in my app, instead of in the .htaccess file they are currently managed in, and came across something I didn't quite understand.

`Symfony\Bundle\FrameworkBundle\Controller\RedirectController` seemed to use dependency injection to get the current request by type-hinting `Symfony\Component\HttpFoundation\Request` for its $request parameter. This seemed odd to me because I had already looked at the list of available services (via console debug:autowiring and debug:container) and didn't see anything wired to this class. The closest thing I saw in this list was the request_stack service, which is what I had been using.

I had seen the case in other frameworks where if the constructor didn't have any dependencies, dependency injection would just instantiate that object and use it. But that's not what's happening because the constructor of the Request class would construct an "empty" version, whereas the variable I am provided through DI is the request that has been created from globals.

**So, how is this request service available even though it doesn't appear to be autowired? Is this something that's just baked into the basic framework code of Symfony?**

Thanks in advance!

**Comment by ahundiak** [comment, marked as answer]
There are actually a number of ways that dependencies get injected into controller action methods.  It is all orchestrated in the HttpKernel::handle method.   If you take a look at the http-kernel/Controller/ArgumentResolver directory you will see a bunch of argument resolver classes including the RequestValueResolver which specifically looks for an argument typehinted to Request and takes care of injecting the current request object.  

It is easy to confuse this sort of injection with what the container does but it is indeed completely independent.  And yes the request handler will instantiate a controller if the controller has not been defined as a service.  This was actually the norm back in the days of 2.x.  It really was not until the autowire functionality was added in Symfony 3.x that defining controllers as services became the norm.

**Reply by jdevinemt** [reply]
Thanks! I'll take a look in that directory.
